### PR TITLE
Assignment6

### PR DESCRIPTION
**Pull request by ionastasi**
Please, review my solution! Any suggestions of how it can be improved?

I have just a few tests, it's better to add more.

### CODE REVIEW

File: assignment6/rearrangingCars_test.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    """
+    moves = rearrangingCars(initial, desired)
+    for m in moves:
+        # we can only move the car to an empty slot
+        if initial[m.destination]:
+            return False
+        initial[m.origin], initial[m.destination] = (initial[m.destination],
+                                                        initial[m.origin])
+    return initial == desired
+
```
Comment (ellapetter) [inline assignment6/rearrangingCars_test.py:21]: Love this idea of testing! :grinning:

**Review by ellapetter — COMMENTED** [review]
Hi,
I really like your solution and the approach using the set instead of iterating over the list as is.
Had nothing major to add or change :)

File: assignment6/rearrangingCars.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        desiredState: list of integers, a permutation of the numbers 0 to N-1.
+                      State of parking lot that we want to get by moving cars.
+        toPrint:      a boolean, True if function should print the sequence of
+                      moves from initalState to desiredState,
+                      and False otherwise.
+
+    Return:
+        list of Move, sequence of moves from state initalState
+        to state desiredState.
+    """
```
Comment (olakalisz) [inline assignment6/rearrangingCars.py:39]: Great documentation!

File: assignment6/rearrangingCars.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    def __eq__(self, other):
+        return (self.origin == other.origin and
+                self.destination == other.destination)
+
+    def __str__(self):
+        return "Move the car from slot {} to slot {}.".format(self.origin,
+                                                            self.destination)
+
+
+def rearrangingCars(initalState, desiredState, toPrint = False):
```
Comment (olakalisz) [inline assignment6/rearrangingCars.py:18]: Very minor issue, the method naming convention in PEP8 suggests using all lowercase letter with underscores in between words [https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables](url)

File: assignment6/rearrangingCars.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        list of Move, sequence of moves from state initalState
+        to state desiredState.
+    """
+
+    slotNums = len(initalState)
+    car2slot = [0] * slotNums
+
+    # create a set of cars that stand on wrong slot, comparing to desiredState
+    # and create map car2slot
+    wrongCars = set()
```
Comment (olakalisz) [inline assignment6/rearrangingCars.py:46]: Suggestion: once the wrong car is found you could immediately  rearrange it and not keep a set ```wrongCars``` in memory 😊 it should save some space

File: assignment6/rearrangingCars.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            # need to return it to the set
+            car = wrongCars.pop()
+            wrongCars.add(car)
+        else:
+            # everything is okey, so we can move a car that should stay at
+            # zeroPos in desiredState
+            car = desiredState[zeroPos]
+            wrongCars.remove(car)
+
+        # move chosen car
```
Comment (olakalisz) [inline assignment6/rearrangingCars.py:74]: Suggestion: Your function ```rearrangingCars``` has is quite long, maybe it would be worth refactoring it and creating the ```move_car``` (or some other name) method with would contain the following piece of code. I think another benefit from doing it would be that you won't have to choose a car ```if desiredState[zeroPos] == 0``` you would simply not call a function in this case. Let me know what you think 😊

File: assignment6/rearrangingCars.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        carPos = car2slot[car]
+        currentState[carPos], currentState[zeroPos] = (currentState[zeroPos],
+                                                        currentState[carPos])
+        car2slot[car], car2slot[0] = car2slot[0], car2slot[car]
+        moves.append(Move(carPos, zeroPos))
+
+    if toPrint:
+        for m in moves:
+            print(m)
+        print("{} moves in total.".format(len(moves)))
```
Comment (olakalisz) [inline assignment6/rearrangingCars.py:84]: Good idea to print the number of moves as well!

File: assignment6/rearrangingCars_test.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    def testEmptyCases(self):
+        self.assertEqual(rearrangingCars([1, 2, 3, 4, 0], [1, 2, 3, 4, 0]), [])
+        self.assertEqual(rearrangingCars([], []), [])
+        self.assertEqual(rearrangingCars([0], [0]), [])
+
+    def testAllCasesWithBruteForce(self):
+        n = 5
+        for inital in permutations(range(n)):
+            for final in permutations(range(n)):
+                self.assertTrue(checkResult(list(inital), list(final)))
```
Comment (olakalisz) [inline assignment6/rearrangingCars_test.py:38]: Very neat tests!

**Review by olakalisz — COMMENTED** [review]
Hi Anastasia! Really good solution, just left a few suggestions 😊

File: assignment6/rearrangingCars.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                      moves from initalState to desiredState,
+                      and False otherwise.
+
+    Return:
+        list of Move, sequence of moves from state initalState
+        to state desiredState.
+    """
+
+    slotNums = len(initalState)
+    car2slot = [0] * slotNums
```
Comment (ellapetter) [inline assignment6/rearrangingCars.py:42]: Maybe rename this? something like cars_position?

File: assignment6/rearrangingCars.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        toPrint:      a boolean, True if function should print the sequence of
+                      moves from initalState to desiredState,
+                      and False otherwise.
+
+    Return:
+        list of Move, sequence of moves from state initalState
+        to state desiredState.
+    """
+
+    slotNums = len(initalState)
```
Comment (ellapetter) [inline assignment6/rearrangingCars.py:41]: Maybe name slotNums as N?
I believe it will match the documentation better :)

File: assignment6/rearrangingCars.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    car2slot = [0] * slotNums
+
+    # create a set of cars that stand on wrong slot, comparing to desiredState
+    # and create map car2slot
+    wrongCars = set()
+    for pos in range(slotNums):
+        # first condition skips 0 because it isn't a car
```
Comment (ellapetter) [inline assignment6/rearrangingCars.py:48]: I know it is the same code wise,
but believe it will be more readable if we have first "car_at_pos = initalState[pos]"
and then refer to "car_at_pos" from now on instead of "initalState[pos]"

File: assignment6/rearrangingCars.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        if initalState[pos] and initalState[pos] != desiredState[pos]:
+            wrongCars.add(initalState[pos])
+        car2slot[initalState[pos]] = pos
+
+    # we don't need to modify initalState, so let's leave it as it is
+    currentState = initalState.copy()
+    moves = list()
+
+    # while we have at least one car that stands on a wrong position
```
Comment (ellapetter) [inline assignment6/rearrangingCars.py:57]: I would rephrase this as "while we have at least one misplaced car "

File: assignment6/rearrangingCars.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    while wrongCars:
+        zeroPos = car2slot[0]
+
+        # choose a car that we will move at this step
+        if desiredState[zeroPos] == 0:
```
Comment (ellapetter) [inline assignment6/rearrangingCars.py:62]: What do you think about switching the order of the two parts? I mean have the "normal" case where desiredState[zeroPos] != 0 and we can just put the right car there, and only afterward address the more special case where desiredState[zeroPos] == 0 where we can just pick up any random misplaced car and place it there?

File: assignment6/rearrangingCars.py (hunk 5 above)
Comment (ellapetter) [reply assignment6/rearrangingCars.py:74]: I think that we will still need to pick a random car in case that desiredState[zeroPos] == 0, because we still have some misplaced cars (since the set is not empty) and we need to have the space somewhere else so we can proceed.. or did I miss something?

**Review by ellapetter — COMMENTED** [review]
Added some more comments :)

File: assignment6/rearrangingCars.py (hunk 4 above)
Comment (ustya-k) [reply assignment6/rearrangingCars.py:46]: (we discussed it earlier, but just in case) @pistacja96 but it saves time. Otherwise she would have to look for a car which is not in it's place if statement on line 62 was correct.

File: assignment6/rearrangingCars.py (hunk 3 above)
Comment (etheodoridou) [inline assignment6/rearrangingCars.py:18]: You are missing an "i" in your input variable: `initialState`

Comment (etheodoridou) [inline assignment6/rearrangingCars.py:18]: Also I would take the `toPrint` variable out and create another function that prints the moves. 

I don't know if you have heard of a set of principles called SOLID. S stands for Single Responsibility Principle. Here is some basic information about it as well: https://en.wikipedia.org/wiki/Single_responsibility_principle

File: assignment6/rearrangingCars.py (hunk 9 above)
Comment (etheodoridou) [reply assignment6/rearrangingCars.py:41]: I would change this to `CAR_PARK_SIZE` as it is more of a constant through your program.

File: assignment6/rearrangingCars.py (hunk 4 above)
Comment (ionastasi) [reply assignment6/rearrangingCars.py:46]: @pistacja96 Also, we already took `O(n)` space when create a map `car2pos`. So it doesn't really metter, in terms of O-notation.

File: assignment6/rearrangingCars.py (hunk 8 above)
Comment (etheodoridou) [reply assignment6/rearrangingCars.py:42]: I suggest changing this to `car_positions` instead

File: assignment6/rearrangingCars.py (hunk 5 above)
Comment (ionastasi) [reply assignment6/rearrangingCars.py:74]: @ellapetter Sure. I do it on line 62 and below.

Comment (ionastasi) [reply assignment6/rearrangingCars.py:74]: @pistacja96 I agree that it's better to crate a new helper function just to make things clearler. But I need this if-statement, in case we have some misplaced car, but an empty space stays where it should.

File: assignment6/rearrangingCars.py (hunk 4 above)
Comment (etheodoridou) [inline assignment6/rearrangingCars.py:46]: I would also suggest changing this to `cars_to_move`, because `wrong_cars` does not explain the purpose of this set

File: assignment6/rearrangingCars.py (hunk 9 above)
Comment (ionastasi) [reply assignment6/rearrangingCars.py:41]: @ellapetter I don't think we should, because `n` is awful name for variable, it says nothing about what it's using for. Also, using one-letter variables is a bad style (except loops).
In documentation I explain the meaning of `n` by words, and docs are quite short. But in code it will be more messy to use `n` instead of `slotNums`.

File: assignment6/rearrangingCars.py (hunk 8 above)
Comment (ionastasi) [reply assignment6/rearrangingCars.py:42]: @ellapetter  Why? Basically, this is the very common way to name maps, `key2value`.

File: assignment6/rearrangingCars.py (hunk 12 above)
Comment (ionastasi) [reply assignment6/rearrangingCars.py:62]: Hmm, may be. I don't think that it makes any difference, but why not, may it will be better then.

File: assignment6/rearrangingCars.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    wrongCars = set()
+    for pos in range(slotNums):
+        # first condition skips 0 because it isn't a car
+        if initalState[pos] and initalState[pos] != desiredState[pos]:
+            wrongCars.add(initalState[pos])
+        car2slot[initalState[pos]] = pos
+
+    # we don't need to modify initalState, so let's leave it as it is
+    currentState = initalState.copy()
+    moves = list()
```
Comment (etheodoridou) [inline assignment6/rearrangingCars.py:55]: I suggest declaring all your variables at the top of your function (after line 42)

File: assignment6/rearrangingCars.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    # and create map car2slot
+    wrongCars = set()
+    for pos in range(slotNums):
+        # first condition skips 0 because it isn't a car
+        if initalState[pos] and initalState[pos] != desiredState[pos]:
+            wrongCars.add(initalState[pos])
+        car2slot[initalState[pos]] = pos
+
+    # we don't need to modify initalState, so let's leave it as it is
+    currentState = initalState.copy()
```
Comment (etheodoridou) [inline assignment6/rearrangingCars.py:54]: I don't think this is necessary. 

I am not sure how good of a practice it is but if you don't use a copy of this, you could also check that the initial state is equal to the final state in your tests, if you wanted to.

File: assignment6/rearrangingCars.py [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    # create a set of cars that stand on wrong slot, comparing to desiredState
+    # and create map car2slot
+    wrongCars = set()
+    for pos in range(slotNums):
+        # first condition skips 0 because it isn't a car
+        if initalState[pos] and initalState[pos] != desiredState[pos]:
+            wrongCars.add(initalState[pos])
+        car2slot[initalState[pos]] = pos
+
+    # we don't need to modify initalState, so let's leave it as it is
```
Comment (etheodoridou) [inline assignment6/rearrangingCars.py:53]: I would delete all the comments inside your code, unless you are explaining something really peculiar you are doing. It just makes your code more confusing in my opinion. I would improve your naming and split your code further to make it more understandable. If you want to explain the general logic of your algorithm, I would do it in your doc string.

File: assignment6/rearrangingCars.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    currentState = initalState.copy()
+    moves = list()
+
+    # while we have at least one car that stands on a wrong position
+    while wrongCars:
+        zeroPos = car2slot[0]
```
Comment (etheodoridou) [inline assignment6/rearrangingCars.py:59]: I suggest you rename this to `empty_slot_pos`

Comment (etheodoridou) [reply assignment6/rearrangingCars.py:59]: I would also add another constant variable to the top: `EMPTY_SLOT = 0`

File: assignment6/rearrangingCars.py (hunk 14 above)
Comment (etheodoridou) [reply assignment6/rearrangingCars.py:54]: I see you used it in your tests, so maybe you shouldn't change it. It's up to you!

File: assignment6/rearrangingCars.py (hunk 5 above)
Comment (etheodoridou) [reply assignment6/rearrangingCars.py:74]: No, that is correct @ellapetter! If you take that condition out, the function turns into an infinite loop

**Review by etheodoridou — COMMENTED** [review]
Here are some comments from me! Mainly for naming and commenting. I have re-written the code myself, just because it helped me review it and test it further. Will post my version of the file on slack, so please have a look :)

File: assignment6/rearrangingCars.py (hunk 8 above)
Comment (etheodoridou) [reply assignment6/rearrangingCars.py:42]: I don't think this is an appropriate convention @Ionastasi. `car2slot` is a bit confusing. This variable just stores the positions of the cars in the car park at any time.

File: assignment6/rearrangingCars.py (hunk 12 above)
Comment (etheodoridou) [reply assignment6/rearrangingCars.py:62]: Isn't it more likely for the position not to be empty? In that case that condition should be checked first :)
